### ISSUE TITLE
Memory consumption is too high and is not freed after processing

### ISSUE THREAD

**Issue by jaimerson** [roo-rb/roo#179]
To process a 2mb file with 16k rows, roo is taking +700mb of memory, and after the file has been processed, the memory isn't freed.
Tested with roo from master, with xlsx and ods formats.

![image](https://cloud.githubusercontent.com/assets/2944985/6513508/40cb92c6-c35a-11e4-9b99-bcbdddec46c9.png)

**Comment by stevendaniels** [comment]
Can you provide a copy of the file along with a sample of the code you used to work with the file?

If you can't provide a copy of the file, can you describe it in more detail? How many columns does it have? What type of data is in the cells, etc. 

Thanks.

**Comment by jaimerson** [comment]
I'm sorry, I can't provide a copy of the file. It has 25 columns and most cell values are decimal values or strings. I'm using ruby version `2.2.0`.

**Comment by jaimerson** [comment]
About the sample of code:
```ruby
def import(file)                                                                                                                            
  spreadsheet = Roo::Spreadsheet.open(file.path, file_warning: :ignore,                                                                                     
                          extension: File.extname(file.original_filename),                                                                      
                          csv_options: { col_sep: ';', converters: :all })                                                                                                                                                
  header = spreadsheet.row(1)                                                                                                               
  (2..spreadsheet.last_row).each do |row|                                                                                                   
     # Some code here                                                                                   
  end                                                                                                                                       
end
```
I tested it and found out that it does not matter what code I pass the block, the significant amount of memory used is the same (even if I only call `next` or do nothing).

**Comment by michaelherold** [comment]
The memory growth is happening during `Roo::Spreadsheet.open`. I'm having the same experience.

The spreadsheet I am using is the Freedom House Freedom of the Press report, available [from Freedom House's website][freedomhouse] ([direct link to file][file]). It's an Excel (.xls) file, which I noticed support being dropped for in the next version of roo.

This isn't a blocking bug for me, as I can just convert the needed part to a CSV and modify it that way, but I thought I'd chime in to possibly help @jaimerson.

Code that causes the memory growth (roo-1.13.2):

```ruby
require 'roo'
require 'uri'  # Needed to prevent a NameError on URI

# Where the xls is a renamed copy of the mentioned sheet ...
sheet = Roo::Spreadsheet.open('freedomhouse-press.xls')
```

[freedomhouse]: https://freedomhouse.org/report-types/freedom-press
[file]: https://freedomhouse.org/sites/default/files/FOTP2014%20Detailed%20Data%20and%20Subscores%201980-2014.xls

**Comment by stevendaniels** [comment]
Thanks. I noticed the issue as well. I've been doing some memory profiling. From my work so far, I haven't noticed any huge wins yet, so far, but I have made some gains. Once my work is done, I'll submit a pull request that will, at the very least, improve things.

**Comment by jaimerson** [comment]
Thank you for your support, guys. Let me know if you need any further information.

**Comment by simonoff** [comment]
@jaimerson It's related to XML what is parsed in memory and to rubyzip what working with archives in memory. What build of nokogiri do you have? With system libraries or with custom patches?

**Comment by simonoff** [comment]
Another possible memory eating it's an formulas and storing data twice - as raw value and as typecasted value

**Comment by stevendaniels** [comment]
@simonoff I'm pretty sure your right, except sometimes it stores data three times: with SharedStrings, it's stored as a sharedString, as a raw value and as a type-casted value. I'm working on some improvements starting with #220 that will address these issues.

**Comment by duncan-bayne** [comment]
I just tried parsing a 27MiB .xlsx file with ~ 1M rows, and Roo ended up consuming > 6.0GiB RAM and locking up my system.  I was going to create a new issue for it, but found this one :smile:

**Comment by simonoff** [comment]
@duncan-bayne what version of Ruby are you using?

**Comment by simonoff** [comment]
@duncan-bayne did you tried streaming reading of file?

**Comment by duncan-bayne** [comment]
@simonoff Ruby 2.2.1.  I hadn't tried streaming the file read - I'll give that a whirl and see if it helps.

**Comment by duncan-bayne** [comment]
Same problem I'm afraid; if I try:

```
 xlsx.each_row_streaming {}
```

... memory usage skyrockets.  I killed Ruby at 5.0GiB, before my system became unresponsive.

**Comment by simonoff** [comment]
hm... can you share this file?

**Comment by stevendaniels** [comment]
@duncan-bayne Could you also tell us how many columns and describe what type of data the cells contain?

When I looked at this issue before, it seemed like Ruby never got around to releasing memory.

**Comment by simonoff** [comment]
Yes, ruby not releasing memory. It keep it for possible usages.

**Comment by smedrick** [comment]
I'm seeing this too when I try to parse an XLSX file that's around 13MB, ~127K rows. Maybe 15 columns, all text.

    xlsx = Roo::Excelx.new(spreadsheet_file_path)
    xlsx.each_row_streaming(pad_cells: true) do |row|
      Rails.logger.info(row.inspect) 
    end

The first row inspect never gets processed. Resque kills the job after 5 minutes trying to process xlsx.each_row_streaming. The server has 2.8GB of RAM.

**Comment by smedrick** [comment]
Well, I feel foolish. Change the code to:

    xlsx.each_row_streaming(max_rows: 1000, offset: count, pad_cells: true) do |row|
      save_row(current_user_id, contact_list_id, row, headers)
    end

with some logic for looping the offset and the code is running fine with only 35% memory usage.

**Comment by benjie** [comment]
I had assumed that `each_row_streaming` was meant to do that for you much like ActiveRecord's `find_each` method - is this not the case?

**Comment by smedrick** [comment]
That was my assumption as well. Turns out the default case is to load the entire file. A bit counterintuitive with that method name.

**Comment by tommyalvarez** [comment]
I identified that also the method 

Roo::Spreadsheet.open(file.path, extension: :xls).last_row for a 250K+ files also spikes memory to more than 2GB. The Roo::Spreadsheet.open(file.path, extension: :xls) does not take much memory, but the last_row to get the total rows of the file does!!

**Comment by tommyalvarez** [comment]
@jaimerson could you take a look to the #last_row method ? That's definitely the cause of high memory consumption for Excels of huge amount of rows. Nearly 5 Gig ram usage for a file with 280K rows

**Comment by jlhonora** [comment]
Any workaround for this issue? Even with `each_row_streaming` the memory usage is insanely high.

**Comment by stevendaniels** [comment]
There are a couple of issues here that need addressing.

1. `last_row` and similar methods are very slow. These methods depend on one of Roo::Excelx::Sheets's hidden features is method called `present_cells`, which filters out empty cells.  This can be very useful if you have a spreadsheet that has many empty rows/columns because the file was saved incorrectly. But in large files, this will make using those methods the first time very, very slow because the entire file must be parsed.

2. `excelx.each_row_streaming` should consume much less memory. I'll need to do some profiling to figure out what exactly is going wrong here.

**Comment by HusseinElMotayam** [comment]
I'm encountering the same issue in my rails application when parsing a 2.7 MB (~3200 rows) .xlsx file.
Inspecting unicorn, the memory usage jumps to around ~700 MB and the memory remains allocated.
`GC.start` is not doing anything.

I got around the problem by:
1. Forking a new process `pid = Process.fork do .... end`.
2. Using roo to read the .xlsx file and push the data to my database. 
3. Extracting some status from IO.pipe in case some parsing problems occur.
4. Waiting for the child process to finish. `Process.wait pid`.

This way, the memory leak is no longer affecting my application as I'm dumping the whole roo context after reading my file.

So, I see unicorn spawning a new process which reaches ~700 MB then terminates. The original unicorn process remains with its memory usage unaffected by the usage of roo and I have my extracted data ready in my database.

Hope this helps.

**Comment by ClayShentrup** [comment]
This is not a problem with garbage collection. Just opening the file uses a huge amount of memory, because `each_row_sctreaming` is keep each row's data in scope the whole time.

**Comment by AlfonsoUceda** [comment]
Hi folks, the problem is here when you use `#each_with_streaming` and under the hood they use this https://github.com/roo-rb/roo/blob/master/lib/roo/excelx/sheet_doc.rb#L161 from https://github.com/roo-rb/roo/blob/master/lib/roo/excelx/sheet_doc.rb#L43, the call to doc loads all the doc so the streaming it isn't streaming.

@stevendaniels can we make a version to avoid process all hyperlinks? maybe one by one?

**Comment by strider9450** [comment]
@jaimerson 
Hi,
Don't you think 2mb is too large a size of an excel with only 16 rows....
Maybe there are thousands of empty rows included in your excel, which are being parsed by Roo and taking so much RAM. Try opening your excel sheet with a text editor and check the amount of stuff in it.

**Comment by jaimerson** [comment]
@strider9450 I know it would be too large if it were 16 rows. I said *16k* rows :wink:

**Comment by igoodrich** [comment]
FWIW, still experiencing the same issue here as well.  I have a 20MB .xlsx file that once loaded in a fresh console with Roo consumes 4.7GB of memory.   Fresh console alone is 300MB.  4.4 GB for a 20MB file is a 400X expansion from disk.  Unsane.

**Comment by SimonBirrell** [comment]
+1 Roo is currently unusable on large (27Mb for us) XLSX files, even with streaming.

**Comment by igoodrich** [comment]
A little more info here.  If I clear formats and save the file it goes from 20MB to 760K and roo has no trouble.  I wonder if there's a way to ignore the formatting in the parsing.

**Comment by rendekarf** [comment]
Hey guys, I have a memory peak of ~700mb for reading a 1.09 MB xlsx file with 45 columns 2.5k rows.

I believe there is something wrong with opening the file and not streaming, as I'm reading the lines in chunks and writing data out to redis before flushing them to the db. When sidekiq starts, it takes about ~120mb of memory but in seconds it spikes to around ~450mb and gradually converges to ~700.

Is there an ongoing work to fix this (in need of help?) or are there any alternative that we can use? This is really annoying.

**Comment by tommyalvarez** [comment]
We ended up convertíng The xlsx file To csv with a Linux system call from
ruby To use a xlsx To csv Linux library. Then I read the csv as usual with
no memory spikes.
This gem won't work well for large excels with rows above ~1000. The
inefficiency of it relies on how it reads all rows and columns and uses
"map" methods against the full collection. Moreover even if you read it in
chunks it won't free memory when you read the next chunk.
The gem clearly needs a refactor for high amount of rows. For us the only
work around was to convert to csv first and use conventional rails
libraries to process csv

**Comment by AlfonsoUceda** [comment]
@tommyalvarez what is the tool you used?

**Comment by pacoguzman** [comment]
Hi guys, one of the things we notice at [bebanjo](https://github.com/bebanjo) is that `Roo::Utils` stores a lot of information between documents parsing and never release that memory.

So we ended doing `Roo::Utils.instance_eval { @split_coordinate = nil }` after each batch of 250 lines while iteration using `each_row_streaming` so we can control a bit more the memory. And releasing again after processing the whole file.

Another improvement is what we did [here](https://github.com/roo-rb/roo/pull/319) maybe you can consider to merge in that PR to decrease the memory consumption, in our case iterating 13000 rows the memory goes from 850MB to 250MB

**Comment by igoodrich** [comment]
FWIW, I have had really good luck reducing memory consumption by selection the entire worksheet and clearing the formatting and then saving.  Huge reduction in size on several occasions now.

**Comment by phillycheeze** [comment]
Also having similar issue with Roo::Spreadsheet.open() and doing `xlsx.row(1).each_with_index`

A 3.1 MB .xlsx file takes up about 1GB of ram in less than two seconds and crashes the process.

**Comment by cyclotron3k** [comment]
Same here. Tried opening a 70MB XLSX; crashed out at > 6GB memory usage.

Now I'm trying to open a 5 MB XLSX, and the CPU usage has been locked at 100% for 10 minutes, and memory usage is creeping past 2 GB. No sign of it completing any time soon.

RUBY_VERSION = "2.1.2"
Roo::Spreadsheet::VERSION = "1.12.1"

**Comment by phillycheeze** [comment]
@cyclotron3k I gave up on trying to get it to work. I would try out [Creek](https://github.com/pythonicrubyist/creek)

It is made for larger XLSX files and consumes much less memory. That 70MB file is going to causes issues, however, even with Creek.

**Comment by cyclotron3k** [comment]
Thanks @phillycheeze. In the end I used a command line tool `ssconvert` to convert the sheets to CSVs, then used the CSV gem. It's an ugly process, but it works. I'll check out creek if I have to revisit the problem.

**Comment by stevendaniels** [comment]
I've merged several PRs that address many of these issues (#332, #333, #334).
Roo releases most of the memory it uses when iterating through a spreadsheet or when using the `each_row_streaming` method. 

There is still some work to do to minimize how and when memory is allocated, but for now, I think this issue can be closed.

**Comment by phillycheeze** [comment]
@stevendaniels thanks for the update. From the benchmarks, it looks like there are some significant memory improvements. Awesome!

**Comment by markmcdonald51** [comment]
Is there some type of remedy to this? I am using roo to try to import a 14mb xlsx file and it locked up an i7 5g w/ 8GB of ram 3 times. This gem kicksazz with files < 10K rows but anything much bigger is starts to choke pretty bad. What are people doing to get around this problem is I may ask?

**Comment by AlfonsoUceda** [comment]
@markmcdonald51 you can give a try to creek gem

**Comment by gabo-mart** [comment]
Is the only option not to use Roo?

**Comment by phillycheeze** [comment]
If you are dealing with large xls or xlsx files, I wouldn't use Roo. Unless you need some markup-specific features (like macros, formulas, styling, etc), your best bet is Creek or using a command line utility like @cyclotron3k suggested. 

The command line utility is a little ugly, but is the best approach memory-wise.

**Comment by stevendaniels** [comment]
re-opening. Will do another deep dive on this.

**Comment by dnagir** [comment]
@stevendaniels just wonder if had a chance to do the deep dive you were planning?
It seems like the the memory leak issue still exists.

**Comment by marisveide** [comment]
Yes... please, can you issue any update?
I have a 5 MB Excel file, where there is obviously something wrong with it, as it has only 20 rows (without k), but there are probably empty cells or something.
The issue is that the client is uploading that file, and I still need to "eat it".
If you could have like col_limit parameter, or something, that would allow to guard against such cases?
Simply Roo::Excelx.new(file_url) eats up like 2+ GB almost instantly, and then gets stuck, and eventually cannibalizes whole RAM.
Excel opens that file without problems.

**Comment by dkkoval** [comment]
If somebody don't need all the awesome features Roo have, you can try [this](https://github.com/dkkoval/fastsheet) gem 😃.

**Comment by andrey-lizunov** [comment]
Any updates connected with memory freed?
Seems that it stack after each file read.
Maybe I should close the stream or something like that?

**Comment by revof11** [comment]
Would this be the same for a call to `sheet.to_csv(target_file)` or should that be a separate ticket?

**Comment by chopraanmol1** [comment]
@revof11 can you provide a sample file for which you are facing the issue? 

Can you also verify if you face a similar issue on roo's master?
